File: LeetCode/146/main.py
class LRUCache:
    class Node:
        def __init__(self, key, val):
            self.key = key
            self.val = val
            self.prev = None
            self.next = None

    def __init__(self, capacity: int):
        self.cap = capacity
        self.head = self.Node(-1, -1)
        self.tail = self.Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.mp = {}

    def add_node(self, newnode):
        temp = self.head.next
        newnode.next = temp
        newnode.prev = self.head
        self.head.next = newnode
        temp.prev = newnode

    def delete_node(self, delnode):
        prevv = delnode.prev
        nextt = delnode.next
        prevv.next = nextt
        nextt.prev = prevv

    def get(self, key: int) -> int:
        if key in self.mp:
            resNode = self.mp[key]
            ans = resNode.val
            del self.mp[key]
            self.delete_node(resNode)
            self.add_node(resNode)
            self.mp[key] = self.head.next
            return ans
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.mp:
            curr = self.mp[key]
            del self.mp[key]
            self.delete_node(curr)

        if len(self.mp) == self.cap:
            del self.mp[self.tail.prev.key]
            self.delete_node(self.tail.prev)

        self.add_node(self.Node(key, value))
        self.mp[key] = self.head.next

File: LeetCode/146/main.cpp
class LRUCache {
public:
    
    class Node {
        public : 
            int key;
            int val;
            Node* next;
            Node*  prev;
        
        Node(int key, int val) {
            this->key = key;
            this->val = val;
        }
    };

    Node* head = new Node(-1, -1);
    Node* tail = new Node(-1, -1);

    unordered_map<int, Node*> mp;
    int cap;

    LRUCache(int capacity) {
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }
    
    void add_node(Node* nn) {
        Node* temp = head->next;
        nn->next = temp;
        nn->prev = head;

        head->next = nn;
        temp->prev = nn;
    }

    void delete_node(Node* nn) {
        nn->next->prev = nn->prev;
        nn->prev->next = nn->next;
    }
    
    int get(int key) {

        if(mp.find(key) != mp.end()) {
            
            Node* nn = mp[key];
            int res = nn->val;

            mp.erase(key);
            delete_node(nn);

            add_node(nn);
            mp[key] = head->next;

            return res;

        }

        return -1;
        
    }
    
    void put(int key, int value) {
        if(mp.find(key) != mp.end()) {
            Node* temp = mp[key];
            mp.erase(key);
            delete_node(temp);
        } else {
            if(mp.size() == cap) {
                Node* lru_node = tail->prev;
                mp.erase(lru_node->key);
                delete_node(lru_node);
            } 
        }
        Node* nn = new Node(key, value);
        add_node(nn);
        mp[key] = head->next;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */